AdyenCard/Components/GiftCardComponent/GiftCardComponent.swift

```diff
@@ -0,0 +1,312 @@
+
+    private func onReadyToPayFullAmount() -> Result<Void, Swift.Error> {
+        AdyenAssertion.assert(message: "readyToSubmitComponentDelegate is nil",
+                              condition: !_isDropIn || readyToSubmitComponentDelegate != nil)
+        stopLoading()
+        if let readyToSubmitComponentDelegate = readyToSubmitComponentDelegate {
+            showConfirmation(delegate: readyToSubmitComponentDelegate)
+        } else {
+            let paymentData = createPaymentData()
+            delegate?.didSubmit(paymentData, from: self)
+        }
+        return .success(())
+    }
+
+    private func showConfirmation(delegate: ReadyToSubmitPaymentComponentDelegate) {
+        let paymentData = createPaymentData()
+        let lastFourDigits = String(numberItem.value.suffix(4))
+        let displayInformation = DisplayInformation(title: "••••\u{00a0}" + lastFourDigits,
+                                                    subtitle: nil,
+                                                    logoName: giftCardPaymentMethod.brand)
+
+        let paymentMethod = PreselectedPaymentMethod(paymentMethod: giftCardPaymentMethod,
+                                                     displayInformation: displayInformation)
+        let component = ReadyToSubmitPaymentComponent(paymentData: paymentData, paymentMethod: paymentMethod)
+        delegate.showConfirmation(for: component)
+    }
+
+    private func requestOrder() -> Result<Void, Swift.Error> {
+        guard let partialPaymentDelegate = partialPaymentDelegate else {
+            AdyenAssertion.assert(message: Error.missingPartialPaymentDelegate.localizedDescription)
+            return .failure(Error.missingPartialPaymentDelegate)
+        }
+        let data = createPaymentData()
+        partialPaymentDelegate.requestOrder(data, from: self) { [weak self] result in
+            self?.handle(orderResult: result)
+        }
+        return .success(())
+    }
+
+    private func handle(orderResult: Result<PartialPaymentOrder, Swift.Error>) {
+        orderResult.handleError { delegate?.didFail(with: $0, from: self) }.handleSuccess(submit(order:))
+    }
+
+    private func submit(order: PartialPaymentOrder) {
+        delegate?.didSubmit(createPaymentData(order: order), from: self)
+    }
+
+    private func createPaymentData(order: PartialPaymentOrder? = nil) -> PaymentComponentData {
+        let details = GiftCardDetails(paymentMethod: giftCardPaymentMethod,
+                                      cardNumber: numberItem.value,
+                                      securityCode: securityCodeItem.value)
+
+        return PaymentComponentData(paymentMethodDetails: details,
+                                    amount: payment?.amount,
+                                    order: order,
+                                    storePaymentMethod: false)
+    }
+}
+
+extension Result {
+
+    @discardableResult
+    func handleSuccess(_ handler: (Success) -> Void) -> Result<Success, Failure> {
```

Comment:
i must say, i'm still not on board with this. feels like result should not be switching on `self`. also it is too easy to forget one of the handlers.

Reply:
its a trad of, your point is valid, but also the benefit of seeing the whole flow in one place is worth it IMHO